velox/connectors/hive/iceberg/PositionalDeleteFileReader.cpp

```diff
@@ -0,0 +1,264 @@
+    uint64_t splitOffset,
+    const std::string& connectorId)
+    : deleteFile_(deleteFile),
+      baseFilePath_(baseFilePath),
+      fileHandleFactory_(fileHandleFactory),
+      executor_(executor),
+      connectorQueryCtx_(connectorQueryCtx),
+      ioStats_(ioStats),
+      pool_(connectorQueryCtx->memoryPool()),
+      filePathColumn_(ICEBERG_DELETE_FILE_PATH_COLUMN()),
+      posColumn_(ICEBERG_DELETE_FILE_POSITIONS_COLUMN()),
+      splitOffset_(splitOffset),
+      deleteSplit_(nullptr),
+      deleteRowReader_(nullptr),
+      deletePositionsOutput_(nullptr),
+      deletePositionsOffset_(0),
+      endOfFile_(false) {
+  VELOX_CHECK_EQ(deleteFile_.content, FileContent::kPositionalDeletes);
+
+  if (deleteFile_.recordCount == 0) {
+    return;
+  }
+
+  // TODO: check if the lowerbounds and upperbounds in deleteFile overlap with
+  //  this batch. If not, no need to proceed.
+
+  std::vector<std::string> deleteColumnNames(
+      {filePathColumn_->name, posColumn_->name});
+  std::vector<std::shared_ptr<const Type>> deleteColumnTypes(
+      {filePathColumn_->type, posColumn_->type});
+  RowTypePtr deleteFileSchema =
+      ROW(std::move(deleteColumnNames), std::move(deleteColumnTypes));
+  auto outputRowType = ROW({posColumn_->name}, {posColumn_->type});
+
+  common::Subfield posSubfield(posColumn_->name);
+  folly::F14FastMap<std::string, std::vector<const common::Subfield*>>
+      outputSubfieldsMap;
+  outputSubfieldsMap[posColumn_->name].push_back(&posSubfield);
+
+  size_t lastPathDelimiterPos = deleteFile_.filePath.find_last_of('/');
+  std::string deleteFileName = deleteFile_.filePath.substr(
+      lastPathDelimiterPos, deleteFile_.filePath.size() - lastPathDelimiterPos);
+  auto deleteTableHandle = std::make_shared<HiveTableHandle>(
+      connectorId, deleteFileName, false, createFilters(), nullptr);
+
+  auto scanSpec = makeScanSpec(
+      outputRowType,
+      outputSubfieldsMap,
+      createFilters(),
+      deleteFileSchema,
+      {},
+      pool_);
+
+  deleteSplit_ = std::make_shared<HiveConnectorSplit>(
+      connectorId,
+      deleteFile_.filePath,
+      deleteFile_.fileFormat,
+      0,
+      deleteFile_.fileSizeInBytes);
+
+  // Create the Reader and RowReader
+
+  dwio::common::ReaderOptions deleteReaderOpts(pool_);
+  configureReaderOptions(
+      deleteReaderOpts,
+      hiveConfig_,
+      connectorQueryCtx_->sessionProperties(),
+      deleteFileSchema,
+      deleteSplit_);
+
+  auto deleteFileHandle =
+      fileHandleFactory_->generate(deleteFile_.filePath).second;
+  auto deleteFileInput = createBufferedInput(
+      *deleteFileHandle,
+      deleteReaderOpts,
+      connectorQueryCtx_,
+      ioStats_,
+      executor_);
+
+  auto deleteReader =
+      dwio::common::getReaderFactory(deleteReaderOpts.getFileFormat())
+          ->createReader(std::move(deleteFileInput), deleteReaderOpts);
+
+  // Check if the whole split can be skipped.
+  if (!testFilters(
```

Comment:
We don't have any data in delete delta file, only path and row numbers, so cannot do this

Reply:
@Yuhta This is to test if the deleteSplit_ on the positional delete file can be skipped (does not need to be read at all). 

- The delete file may not contain the positions of the base file IcebergSplitReader is reading, because an Iceberg positional delete file may contain positions from multiple base files. For example, suppose we are reading base_file_2, and the positional delete contains the following rows
```
path_to_base_file_ 1, 100
path_to_base_file_ 1, 105
path_to_base_file_ 3, 1
```
This commits already has the path column filter added.

-  The delete file may not contain the positions in the current batch that is being read by IcebergSplitReader. Suppose the IcebergSplitReader wants to read rows in [0, 10000] in the base file in the current batch, but the positional delete file does not contain the rows in this range, then the delete file doesn't need to be read. i.e. delete file has the following rows
```
path_to_base_file_ 1, 50000
path_to_base_file_ 1, 50010
```
This filter will be added soon in future commits.